### ISSUE TITLE
FR: Show book length/page count for all books

### ISSUE THREAD

**Issue by rixx** [koreader/koreader#8390]
The only thing I'm really missing from the Kindle book browser is their indicator of rough book length (with a dotted line) in the list of books. I'll often choose the next book by rough feeling for the length, and Koreader only shows the page count once a book has been opened.

Ideally I'd also like to be able to sort by page count, but just having it displayed, visually or as number, would be good enough! I realise that file size is shown, but file size is not a great indicator for book length (and in fact, 

I assume that this isn't currently done because getting the page count for all books is potentially slow and blocking, more so than the first time load of book covers? Could there be an option to trigger page count manually, or to supply it via an additional file (e.g. automatically from Calibre)?

**Comment by NiLuJe** [comment]
Long-press on a folder or on the plus icon in the top right has an option to scan all the things. IIRC, that will include page-count (or not, can't remember).

**Comment by rixx** [comment]
It might extract the page count, but it definitely doesn't show it unless the book has been opened before.

**Comment by mergen3107** [comment]
No, the scan thing only extracts metadata and covers for the current view. Pages are only extracted once you open it — because it needs to know what settings are you using like fonts, sizes, etc. 

It would be interesting to be able able to scan books for pages using default settings, but this is not a feature. 

Besides, the page count will become irrelevant once you change default settings.

**Comment by rixx** [comment]
That's a good point, and might be why the default Kindle view does not give a page count – which is not particularly interesting, after all.

But some indicator to tell long books apart from short books would be really nice – I don't really care if the underlying metric is page count, page count in default layout, word count, or something else – pretty much anything that counts *something* in the books is going to be better than file size.

**Comment by Fank1** [comment]
I'm totally with you here @rixx . It was one of the best features on the Kindle, which they removed. Since I switched to **Kobo**, and thereafter installed **Koreader** – I have missed (and yearned for) it. ;) Why is it so good? 

1. It quickly gives you an indication on how long the book compared to your other books in the library. 
2. It doubles as a reading progress indicator

Technically (as I remember it) the dot length was not relative to the other books in the library, and not relative to your font size (which would alter the total pages as you guys have mentioned). They had some way of determining book length that made for example _Cloud Atlas_ (see image) always 60 dots. 

Anyway, I would so much support the building of this feature. Sadly, I don't know Lua. I work as a UI designer and would love to contribute any way I can (even though I understand that the UI design is not the difficult part). 

![kindle](https://user-images.githubusercontent.com/19385327/139591835-7ee53a81-f22e-4a85-9d03-355236be15db.png)

**Comment by Frenzie** [comment]
> Technically (as I remember it) the dot length was not relative to the other books in the library, and not relative to your font size (which would alter the total pages as you guys have mentioned). They had some way of determining book length that made for example Cloud Atlas (see image) always 60 dots.

There are two methods I can think of:

1. A dumb but simple method: just see how big the XHTML files are in the books.
2. A better way, same as above but with all the XHTML markup stripped so it's a text-only output that's mostly accurate. But I doubt anybody's going to bother writing something like that.

On that basis you could then say that every 5000 words, every 25 000 characters, or something vaguely like that, is one mark.

**Comment by offset-torque** [comment]
**An idea:**
There is a Calibre plugin that count pages of books (according to different algorithms) and write into book as custom metadata (As far as I understand). Maybe KOReader can show that number if it finds the count within metadata.
[Calibre Count Pages Plugin](https://www.mobileread.com/forums/showthread.php?t=134000)

If a user wants page counts, he can count them within Calibre before transferring to device. Other users won't be affected and KOReader doesn't need to deal with counting.

**Comment by christian-ho** [comment]
I also think it would be a great idea if you could specify in the file list which meta data fields are displayed.
For example, I don't need the file size or the file format, instead I would prefer to show the word count from Calibre.
But I also know that this is probably not so easy?
![FileManager_2021-11-20_172321](https://user-images.githubusercontent.com/57492542/142733849-fbc00bba-b1ac-4f58-8b93-1d316e67c252.jpg)

**Comment by scsmartt** [comment]
Would love to have this feature as well.

**Comment by benoit-pierre** [comment]
Here is the patch to also calculate the number of pages when extracting a CRE document information:
```diff
---
 plugins/coverbrowser.koplugin/bookinfomanager.lua | 16 ++++------------
 1 file changed, 4 insertions(+), 12 deletions(-)

diff --git a/plugins/coverbrowser.koplugin/bookinfomanager.lua b/plugins/coverbrowser.koplugin/bookinfomanager.lua
index 003f59eb..fce21ba1 100644
--- a/plugins/coverbrowser.koplugin/bookinfomanager.lua
+++ b/plugins/coverbrowser.koplugin/bookinfomanager.lua
@@ -473,24 +473,16 @@ function BookInfoManager:extractBookInfo(filepath, cover_specs)
     local document = DocumentRegistry:openDocument(filepath)
     local loaded = true
     if document then
-        local pages
         if document.loadDocument then -- needed for crengine
-            if not document:loadDocument(false) then -- load only metadata
+            if document:loadDocument() then
+                document:render()
+            else
                 -- failed loading, calling other methods would segfault
                 loaded = false
             end
-            -- For CreDocument, we would need to call document:render()
-            -- to get nb of pages, but the nb obtained by simply calling
-            -- here document:getPageCount() is wrong, often 2 to 3 times
-            -- the nb of pages we see when opening the document (may be
-            -- some other cre settings should be applied before calling
-            -- render() ?)
-        else
-            -- for all others than crengine, we seem to get an accurate nb of pages
-            pages = document:getPageCount()
         end
         if loaded then
-            dbrow.pages = pages
+            dbrow.pages = document:getPageCount()
             local props = document:getProps()
             if next(props) then -- there's at least one item
                 dbrow.has_meta = 'Y'
```
Note:
* this, of course, makes the process slower
* on master, the "Book information" dialog does not currently take that information into account (broken since b0eb0ce0, see #9647)
* there does not seem to be a way to easily apply all the reader default settings, so the count will differ from the value gotten after opening a book

And for a graphical length / progress indicator:
```diff
---
 plugins/coverbrowser.koplugin/listmenu.lua | 25 ++++++++++-
 plugins/coverbrowser.koplugin/main.lua     | 68 ++++++++++++++++++++++++++++++
 2 files changed, 92 insertions(+), 1 deletion(-)

diff --git a/plugins/coverbrowser.koplugin/listmenu.lua b/plugins/coverbrowser.koplugin/listmenu.lua
index 7db9cedb..cb6fc491 100644
--- a/plugins/coverbrowser.koplugin/listmenu.lua
+++ b/plugins/coverbrowser.koplugin/listmenu.lua
@@ -37,6 +37,16 @@ local getMenuText = require("ui/widget/menu").getMenuText
 
 local BookInfoManager = require("bookinfomanager")
 
+-- Available status display modes
+local STATUS_DISPLAY_MODES = {
+    -- nil or ""                -- detailed page count and progress status
+    graphical_1 = {'⬥', '⬦'},   -- graphical: ⬥⬥⬦⬦⬦
+    graphical_2 = {'⬧', '⬨'},   -- graphical: ⬧⬧⬨⬨⬨
+    graphical_3 = {'⦁', '⚬'},   -- graphical: ⦁⦁⚬⚬⚬
+    graphical_4 = {'◾', '◽'}, -- graphical: ◾◾◽◽◽
+    graphical_5 = {'⬝', '⬞'},   -- graphical: ⬝⬝⬞⬞⬞
+}
+
 -- Here is the specific UI implementation for "list" display modes
 -- (see covermenu.lua for the generic code)
 
@@ -441,6 +451,17 @@ function ListMenuItem:update()
                 end
             end
 
+            local status_display_mode = BookInfoManager:getSetting("status_display_mode")
+            if bookinfo.pages ~= nil and status_display_mode then
+              local num_chars = math.floor((bookinfo.pages + 99) * (1/100))
+              local completed = percent_finished and math.floor(percent_finished * num_chars) or 0
+              local indicators = STATUS_DISPLAY_MODES[status_display_mode];
+              pages_str = string.rep(indicators[1], completed) .. string.rep(indicators[2], num_chars - completed)
+              if status == "complete" or status == "abandoned" then
+                pages_str = (status == "complete" and _("Finished") or _("On hold")) .. " – " .. pages_str
+              end
+            end
+
             -- Build the right widget
 
             local fontsize_info = _fontSize(14, 18)
@@ -878,7 +899,9 @@ end
 
 -- Simple holder of methods that will replace those
 -- in the real Menu class or instance
-local ListMenu = {}
+local ListMenu = {
+  STATUS_DISPLAY_MODES = STATUS_DISPLAY_MODES,
+}
 
 function ListMenu:_recalculateDimen()
     -- Find out available height from other UI elements made in Menu
diff --git a/plugins/coverbrowser.koplugin/main.lua b/plugins/coverbrowser.koplugin/main.lua
index f030a2fb..47120ded 100644
--- a/plugins/coverbrowser.koplugin/main.lua
+++ b/plugins/coverbrowser.koplugin/main.lua
@@ -44,6 +44,7 @@ local filemanager_display_mode = false -- not initialized yet
 local history_display_mode = false -- not initialized yet
 local collection_display_mode = false -- not initialized yet
 local series_mode = nil -- defaults to not display series
+local status_display_mode = nil -- defaults to detailed
 
 local CoverBrowser = WidgetContainer:extend{
     name = "coverbrowser",
@@ -72,6 +73,7 @@ function CoverBrowser:init()
     self:setupHistoryDisplayMode(BookInfoManager:getSetting("history_display_mode"))
     self:setupCollectionDisplayMode(BookInfoManager:getSetting("collection_display_mode"))
     series_mode = BookInfoManager:getSetting("series_mode")
+    status_display_mode = BookInfoManager:getSetting("status_display_mode")
 
     init_done = true
     BookInfoManager:closeDbConnection() -- will be re-opened if needed
@@ -227,6 +229,53 @@ function CoverBrowser:addToMainMenu(menu_items)
                     },
                 },
             },
+            {
+                text = _("Detailed list: status display mode"),
+                sub_item_table = {
+                    {
+                        text = _("Detailed: page count & progress status"),
+                        checked_func = function() return not status_display_mode end,
+                        callback = function()
+                            self:setupStatusDisplayMode()
+                        end,
+                    },
+                    {
+                        text = _("Graphical: ⬥⬥⬦⬦⬦"),
+                        checked_func = function() return status_display_mode == 'graphical_1' end,
+                        callback = function()
+                            self:setupStatusDisplayMode('graphical_1')
+                        end,
+                    },
+                    {
+                        text = _("Graphical: ⬧⬧⬨⬨⬨"),
+                        checked_func = function() return status_display_mode == 'graphical_2' end,
+                        callback = function()
+                            self:setupStatusDisplayMode('graphical_2')
+                        end,
+                    },
+                    {
+                        text = _("Graphical: ⦁⦁⚬⚬⚬"),
+                        checked_func = function() return status_display_mode == 'graphical_3' end,
+                        callback = function()
+                            self:setupStatusDisplayMode('graphical_3')
+                        end,
+                    },
+                    {
+                        text = _("Graphical: ◾◾◽◽◽"),
+                        checked_func = function() return status_display_mode == 'graphical_4' end,
+                        callback = function()
+                            self:setupStatusDisplayMode('graphical_4')
+                        end,
+                    },
+                    {
+                        text = _("Graphical: ⬝⬝⬞⬞⬞"),
+                        checked_func = function() return status_display_mode == 'graphical_5' end,
+                        callback = function()
+                            self:setupStatusDisplayMode('graphical_5')
+                        end,
+                    },
+                },
+            },
         },
     }
 
@@ -735,4 +784,23 @@ function CoverBrowser:setupCollectionDisplayMode(display_mode)
     end
 end
 
+function CoverBrowser:setupStatusDisplayMode(display_mode)
+    local ListMenu = require("listmenu")
+    if not ListMenu.STATUS_DISPLAY_MODES[display_mode] then
+        display_mode = nil -- unknow mode, fallback to detailed
+    end
+    if init_done and display_mode == status_display_mode then -- no change
+        return
+    end
+    if init_done then -- save new mode in db
+        BookInfoManager:saveSetting("status_display_mode", display_mode)
+    end
+    -- remember current mode in module variable
+    status_display_mode = display_mode
+    logger.dbg("CoverBrowser: setting status display mode to:", display_mode)
+    if string.find(filemanager_display_mode, '^list_') then
+        self:refreshFileManagerInstance()
+    end
+end
+
 return CoverBrowser
```

Screenshots:
![2022-10-17-215623_540x720_scrot](https://user-images.githubusercontent.com/5104286/196272189-f1496ad5-6558-43df-8277-68f1ee795744.png)
![2022-10-17-215652_540x720_scrot](https://user-images.githubusercontent.com/5104286/196272204-138c51c7-e5d4-4a6c-9510-971f597b1ab1.png)

**Comment by hius07** [comment]
I believe these graphical dots are good for rough visual estimation of the book length / progress.
Not sure of a setting to choose the type, just small/big dots like Kindle shows are good for me.
I would add a setting of number of pages per dot.

**Comment by offset-torque** [comment]
> **hius07:** I would add a setting of number of pages per dot

+1 for this ^^^

Also is it possible to have more space between the dots ? 
In Kindle for example, dots are **2px** and spaces between them are **4px**.
This gives visual separation between the dots even though they are smaller.

In the proposal, dots are literally touching. They may look nice in big, sharp screenshots but in real life with small e-ink screens under non-ideal lighting conditions, they can lose distinction and blur into a regular progress bar. Visual separation between elements enhances glanceability and (imho) looks more pleasing. 

**In this mockup I just copy pasted your circles:**

![book length dots example](https://user-images.githubusercontent.com/84420582/196344911-e9afffe6-2971-4b4e-8f7a-a2d13ad96f90.png)

And I think using diamonds is a good idea. It can provide even more distinction between the dots.

**Comment by poire-z** [comment]
No strong opinion about adding an alternative to `P % of N pages` with dots/diamonds - but if we do, only a single type of them (adding multiple choices is probably excessive, but nice to propose so we can choose and agree on one).
You would still need to do something with overflow when you meet a book with 8000 pages - and/or live with the single dot/diamond for books of 60 vs 120 pages. (May be some math lovers among use can put out a logasomethingrithmic formula that could make this nicer.)

> this, of course, makes the process slower
```diff
-            if not document:loadDocument(false) then -- load only metadata
+            if document:loadDocument() then
+                document:render()
+            else
```

A LOT slower. Currently, we load only metadata, so it takes 1s or 2s per book. By loading the XML and rendering, you will actually open the book. Your War and Peace will take 40s. Browsing your Tolstoi complete works directory will take 5 minutes. And then browsing your Dostoievski and Proust subdirectories will get your device's battery empty.
And moreover, the innacurate number of pages you'll get for such never-opened books won't relate to the accurate number of pages you have for already-opened books.
Soft veto on even adding this as an option: users would probably enable it because why not - but browsing your library would be slow, unaccurate, and not a good user experience.

**Comment by NiLuJe** [comment]
I wholeheartedly agree with everything @poire-z said.

(I, err, don't even *see* the interest of getting a pagecount that'll be inaccurate anyway).

**Comment by offset-torque** [comment]
> NiLuJe: I, err, don't even see the interest of getting a pagecount that'll be inaccurate anyway

**Why don't we do it manually ?**
As I see it, every type of automatic calculation on device is costly (cpu time/battery) and inaccurate because page number calculation depends on many factors.

**Imagined solution:**
1. User transfers the book to their device and opens it
2. Looks up the book from Amazon, Goodreads etc. and determine page number
3. Opens the book info dialog (or smt like that) and enters the page number
4. File browser shows that number

**Benefits:**
- No extra processing
- Page number is accurate
- Doesn't cost anything to users who don't care about this

**What this solves ?**
From the Owner of the issue:

> - I'll often choose the next book by rough feeling for the length, and Koreader only shows the page count once a book has been opened.
> 
> - Ideally I'd also like to be able to sort by page count, but just having it displayed, visually or as number, would be good enough!

And it is not a big overhead. Even if the user reads one book a day, looking up the page number and entering it takes like 30 seconds. So why not offload a bit of work to the user to simplify the process enormously.

**Comment by offset-torque** [comment]
Did you even read the text you quoted @pazos 
Let me summarize:
- OP asked for a way to see page count.
- I offered a solution to enter the page count.

> The patch above is cool. 

So this is your definition of cool:

> **poire-z:** Your War and Peace will take 40s. Browsing your Tolstoi complete works directory will take 5 minutes. And then browsing your Dostoievski and Proust subdirectories will get your device's battery empty.

What will you say to the users ?
**"Here is a cool patch. But don't click on directories with many books because it kills your device."**
Any device which acts like this is fundamentally broken. 
Here, I found your bazooka.

**Comment by poire-z** [comment]
> And it is not a big overhead. Even if the user reads one book a day, looking up the page number and entering it takes like 30 seconds. So why not offload a bit of work to the user to simplify the process enormously.

Well, we can even make his life easier: no need to lookup on Amazon and input the number. Just open that book in KOReader :) will take a few seconds and rarely minutes.

I understand the request, and me too, sometimes, when having to decide my next book, I'd like to know their size so I can decide to go with a 100 pages book instead of a 800 pages book when I know I'll soon get busy. It's a bit bothering opening a new book for me (I honestly hesitate each time :) as I'll have to then reset book settings as I like my directories tidy. But I'm sure lots of users just do that with newly added books.

But some users would like to see this for the 200+ books library, and opening the books one by one is bothering.
The solution feels it should be CoverBrowser's bookinfo module job to do more work and compute the page numbers - but no, it's just there to quickly fetch and show metadata, while allowing browsing directories fast. The page numbers are not in the metadata and not static, and expensive to compute. So we just don't do it and we all have to live with it.

**Comment by offset-torque** [comment]
I think there is a misunderstanding.
I didn't suggest replacing KOReader book count with the user entered data.
I meant this as a new field maybe called **"Physical Page Count"** which is separate from the KOReader calculated one.
And if a user wants to see this, it can be shown like the file size in the **File Browser**.

**Comment by poire-z** [comment]
And various random technical thoughts:

There could be a user-plugin that would navigate all your library and open each non-opened document, that you could launch with your device plugged before going to bed :)

If doing it by patching (optional setting or user-patching) `BookInfoManager:extractBookInfo()` like it was proposed above, the number will be innacurate if we don't use all the settings (and screen size, accounting for little things like footer header, ...).
Duplicating the default settings setup in this function is not a solution: it would add kludge and we'll forget to keep it updated.
The real solution would be to have BookInfoManager:extractBookInfo() actually have ReaderUI open the book in the background, with the same code involved when really opening it manually. And for that, it would need to have/make fake UIManager, Screen, Device... overriding their real functions with dummy/no-op ones so it doesn't effect the main thread device state. Which is super complex :) and will still take 40s - but would give the accurate page numbers.

I'm not fond of any other solution (less complex than the above) that would give innacurate some pages number (like just doing the render() above without any setting).

One still innacurate solution that crossed my mind:
`document:loadDocument(false)` could be made to sum the sizes of all the .xhtml files in an EPUB that would be parsed when opening, and return the total size of the raw HTML.
This could be gathered from opened books too, and we could compute some statistics about computed rendered nb of pages with the user usual settings / total size of the raw HTML, and use this user-specific ratio to compute some estimated pages number for a book when we can only get that total size of the raw HTML.

**Comment by poire-z** [comment]
> I meant this as a new field maybe called "Physical Page Count" which is separate from the KOReader calculated one.

I understand, but I'm not sure many people will keep that updated. May be people using Calibre can keep their computer library tidy with all this info, and maybe it's forwarded to the data on the device and used/usable by our calibre plugin.

Also, about adding this kind of fields: we don't have a database to store such out-of-band information. The only information we have is stored:
- in the book itself, its metadata
- in an opened book docsettings (its bookname.sdr/metadata.epub.lua)

We don't have another permanent store for other information (except may be the calibre plugin that I know nothing about). Oh, and also the reading statistics plugin stats.db.

CoverBrowser's own database is just a cache of the book metadata. It can be trashed (when its schema is updated or at the user wills) and rebuilt. I don't want to have it be valuable with real data that would make its management more complex.
So, no place to store a publisher hardcover number of pages.
(We don't even have a solution to edit title/authors when the metadata is crap or messed up :) and if we did, this editing would be volatile, and disappear on the next cache rebuild. The solution is to edit the book and its metadata on a computer.)

**Comment by hius07** [comment]
Calculating of the number of pages can be fired manually via a button in the long-press popup menu. It can be done for a group of selected files (and we have a button Select all files in a folder). User would know that it takes some time.
Where to store? In a new entry in doc_settings in the sdr folder. We would need some changes in determining of a book status that has not been really opened yet but has a sdr folder.

**Comment by pazos** [comment]
@offset-torque: your suggestion would be as cool as the patch above if somebody implements it and makes an user patch.

What I think about user patches?

Of course [there's an xkcd](https://xkcd.com/763/) for that
![workaround](https://user-images.githubusercontent.com/975883/196527664-f93a2916-ae0d-4368-bb2b-6696fbf4112c.png)

Does it mean all user patches look like a good idea? **Nooooooo**. But hey, there's the bazooka, have fun.

**Comment by Frenzie** [comment]
@poire-z 
> One still innacurate solution that crossed my mind:
document:loadDocument(false) could be made to sum the sizes of all the .xhtml files in an EPUB that would be parsed when opening, and return the total size of the raw HTML.
This could be gathered from opened books too, and we could compute some statistics about computed rendered nb of pages with the user usual settings / total size of the raw HTML, and use this user-specific ratio to compute some estimated pages number for a book when we can only get that total size of the raw HTML.

Yup, that's also the only method I see as viable on-device. Cf. https://github.com/koreader/koreader/issues/8390#issuecomment-955742151

**Comment by NiLuJe** [comment]
> One still innacurate solution that crossed my mind:
> `document:loadDocument(false)` could be made to sum the sizes of all the .xhtml files in an EPUB that would be parsed when opening, and return the total size of the raw HTML.
> This could be gathered from opened books too, and we could compute some statistics about computed rendered nb of pages with the user usual settings / total size of the raw HTML, and use this user-specific ratio to compute some estimated pages number for a book when we can only get that total size of the raw HTML.

Fun fact: that's (very) roughly how the original "position" implementation worked for MOBI on Kindle.
And roughly how Adobe's RMSDK page numbers work, too.

(And you just have to take a dip into MobileRead to see the kind of passions this stupid argument can trigger :D).

**Comment by poire-z** [comment]
> Calculating of the number of pages can be fired manually via a button in the long-press popup menu. It can be done for a group of selected files (and we have a button Select all files in a folder). User would know that it takes some time.

It would still be some huge work on the code to render correctly in the background - all that just to get accurate page numbers :/
Unless, as the user requests it, you just open (in the main thread, so he will visually notice all these openings) all the selected books. Then, just keep the regular docsettings (or purge them from everything except the number of pages and the metadata).
Still ugly.

> Where to store? In a new entry in doc_settings in the sdr folder. We would need some changes in determining of a book status that has not been really opened yet but has a sdr folder.

There's a years old covenant that .sdr/metadata.xyz.lua means opened - no .sdr means never opened. I quite like it: when mounting the filesytem on my computer, I know what I have opened and decided to keep (for some rare highlights I do). I'm not ready to sacrifice this "feature", so I'll fight to keep the old traditions :)

> Fun fact: that's (very) roughly how the original "position" implementation worked for MOBI on Kindle.
And roughly how Adobe's RMSDK page numbers work, too.
(And you just have to take a dip into MobileRead to see the kind of mess this stupid argument can unleash :D).

These mobileread threads are mostly about page numbering, real vs screen vs fake.
(We could mostly propose them all, it just needs some little work I have started but that I procrastinate ending... https://github.com/koreader/koreader/issues/9020#issuecomment-1119588295 .)
Do these fake page numbering methods really makes out the total nb of page displayed in their file browsers ? And do they display it before the book is opened ? Or do they get it via some other metadata (so, matching the page numbering as it has been computed the same way upstream)?

**Comment by NiLuJe** [comment]
> Do these fake page numbering methods really makes out the total nb of page displayed in their file browsers ? And do they display it before the book is opened ? Or do they get it via some other metadata (so, matching the page numbering as it has been computed the same way upstream)?

They used to. Nowadays there's often some more crap happening, either heuristically or based on magic metadata values coming from god know where ;).

**Comment by NiLuJe** [comment]
The issue inherent to these approaches (besides the obvious), is that the results will be even more heavily skewed for markup-heavy files (of which there are... unfortunately many). An obvious example would be a KePub, because of the added spans ;).

I have no idea if you can actually get the *text* length only from CRe, because that would be much, much, much better ;).

**Comment by poire-z** [comment]
> I have no idea if you can actually get the text length only from CRe, because that would be much, much, much better 

We could, but I would prefer not to have to read and parse all the HTML to get only text content.

**Comment by benoit-pierre** [comment]
That's what `buildSyntheticPageMapIfNoneDocumentProvided` does. Using that function is actually slower then rendering for some reason (something to do with validating the generated page map later), and does not always complete successfully. I added a new `estimatePageCount` function, and that one is faster.

Refreshing a list of 25 books on my aging Kindle Paperwhite, the timings are:
* 17 seconds with the original code (metadata only page count extraction)
* 127 seconds (x8) when rendering to estimate the page count
* and 49 seconds (x3) with the new `estimatePageCount` implementation
* using `buildSyntheticPageMapIfNoneDocumentProvidedOrig`/`getPageMapLastPageLabel` original implementation is much slower and does not complete successfully on some of the bigger files: 435 seconds (x25)

Updated patches [here](https://github.com/koreader/koreader/compare/master...benoit-pierre:koreader:bpierre), [here](https://github.com/koreader/koreader-base/compare/master...benoit-pierre:koreader-base:bpierre) and [here](https://github.com/koreader/crengine/compare/master...benoit-pierre:crengine:bpierre).

**Comment by benoit-pierre** [comment]
Currently, `BookInfoManager:extractBookInfo` always re-extract everything, wouldn't it make more sense to only extract missing information, except when asked to explicitly refresh everything?

Additionally, would it be possible to change the sequence of confirm dialogs (`Here only?`, `Refresh?`, `Prune`, …) to a single dialog with a series of checkboxes (saving the state of those for future invocations)?

**Comment by poire-z** [comment]
> Currently, BookInfoManager:extractBookInfo always re-extract everything, wouldn't it make more sense to only extract missing information, except when asked to explicitly refresh everything?

What 2+ sets of different into would make "everything" ? What is "missing information" and how ? Because it gets ALL the metadata from the engine at once, and nothing else from elsewhere - so I don't really get what you're thinking about.

> Additionally, would it be possible to change the sequence of confirm dialogs (Here only?, Refresh?, Prune, …) to a single dialog with a series of checkboxes

The way this single function works is by using our Trapper module, which supports currently only confirm() and info(), and allows to have it linear without nested callbacks.
So, possible: yes. Complex: surely. Worth it: probably not, this is seldom used, a normal use should be fine with seeing these few consecutive confirmboxes every other month.

**Comment by benoit-pierre** [comment]
> > Currently, BookInfoManager:extractBookInfo always re-extract everything, wouldn't it make more sense to only extract missing information, except when asked to explicitly refresh everything?
> 
> What 2+ sets of different into would make "everything" ? What is "missing information" and how ? Because it gets ALL the metadata from the engine at once, and nothing else from elsewhere - so I don't really get what you're thinking about.

Which is fine when it's only metadata but is problematic if some if this information is costly to compute (e.g. estimated page count), no?

**Comment by poire-z** [comment]
> Which is fine when it's only metadata but is problematic if some of this information is costly to compute (e.g. estimated page count), no?

But when would you need to get only one of these 2 refreshed ?
The metadata are static and cheap, the page numbers (in case you change your fetching method) are expensive.
But a user rarely/never change the metadata (can only be done on a computer, and for us, no garantee the book hasn't changed content and page numbers), so your only (developper) use case is refreshing the page numbers. Refetching metadata costs so-nothing more in this context.

**Comment by benoit-pierre** [comment]
I thought updating the covers would trigger it (e.g. switching from the list view to the mosaic), but looks like I was wrong, never mind.

**Comment by poire-z** [comment]
> I thought updating the covers would trigger it (e.g. switching from the list view to the mosaic), but looks like I was wrong, never mind.

Not that wrong :) this can happen when switching from list to mosaic, when only small covers (for small list items) were cached: we re-do the extraction to get the original cover and cache larger covers so they look nice on mosaic items). (Then the other way around mosaic>list won't trigger it - we can use the larger mosaic cached covers to display them scaled down on list items - and then list>mosaic: the larger mosaic are still there - so probably why you couldn't reproduce :)
Again, if fetching the covers (and rescaling), getting the metadata is nearly free.

**Comment by poire-z** [comment]
> That's what `buildSyntheticPageMapIfNoneDocumentProvided` does. Using that function is actually slower then rendering for some reason (something to do with validating the generated page map later), and does not always complete successfully.

To use this function, you actually have to load the document (parse the XHTML, build the DOM - and as a shortcut done in this case: apply CSS to each node). You can see the duration of this phase when you open a book: it's the first half of the opening progress bar at top left.
The rendering (measuring paragraphs, making pages) is the second half of this progress bar.
So, roughly, loading time = rendering time = each half the total opening time.

buildSyntheticPageMapIfNoneDocumentProvided() itself makes a list of xpointers, and may be it tries to resolve these xpointers to rendered page numbers - and may be this is messed up on a non-rendered document. Anyway, probably not worth investigating.

Also, using it, you will get the number of synthetic pages: they will relate to each other non-opened documents, but won't relate to opened document's page numbers.
Users would then get the habit of multiplying that number by some personal ratio (_it shows 120 pages, I know I need x2 as experience has teached me **I** will get 240 rendered pages_).
You could have them tweak the 1024 chars per synthetic pages, but well...
Or you could do as I tested above to compute a ratio on opened books: rendered pages / synthetic pages.
Anyway, still feels expensive, mostly because of the uneeded DOM building.
The proper way to do this - and I have no idea of the duration decrease it would bring and if it's worth truing, would be to do, in epubfmt.cpp and in all the other formating codes (they all do the same thing):
- use the existing XMLParser/HTMLParser as they are used, to parse each document fragment
- instead of having them feed a ldomDocumentWriter (onTagOpen, onText), have them feeld a ldomDocumentCharCounter, that would just count the size of text given to OnText) - so no DOM and more expensive stuff is done.

But would all this really be worth the time (developper and device time) to get something more precise, when it will still be innacurate/not-relatable.
For the purpose of the issue (or at least my appreciation of its need), anything cheap and rough is good enough.

> Refreshing a list of 25 books on my aging Kindle Paperwhite, the timings are:
    17 seconds with the original code (metadata only page count extraction)
    127 seconds (x8) when rendering to estimate the page count
    and 49 seconds (x3) with the new estimatePageCount implementation

I'm surprised you only get a 17>49 (also surprised by this 17>127).
I think it's this 17 that may be surevaluated - I think it should take a few 100ms per book. May be it is made larger by the eInk refreshes showing progress (with batch extraction) or the round trips in a as-you-browse extraction - which are the real use cases and timings, so not using that as an argument :)
But still, anything more than 10 seconds to see a whole page info extracted feels excessive.

(There could be some optimisations done when looking for files by name in the zip container: currently, if 100 files, when looking for one file by name, we will iterate all until we find it, doing each time a loooong string comparison: this is slow, and probably the reason when a book metadata refresh seems to take 4s while an other takes 1s - its EPUB must contain hundreds of files...).

**Comment by benoit-pierre** [comment]
> Also, using it, you will get the number of synthetic pages: they will relate to each other non-opened documents, but won't relate to opened document's page numbers.

I know.

> > Refreshing a list of 25 books on my aging Kindle Paperwhite, the timings are:
> > 17 seconds with the original code (metadata only page count extraction)
> > 127 seconds (x8) when rendering to estimate the page count
> > and 49 seconds (x3) with the new estimatePageCount implementation
> 
> I'm surprised you only get a 17>49 (also surprised by this 17>127). I think it's this 17 that may be surevaluated - I think it should take a few 100ms per book. May be it is made larger by the eInk refreshes showing progress (with batch extraction) or the round trips in a as-you-browse extraction - which are the real use cases and timings, so not using that as an argument :) But still, anything more than 10 seconds to see a whole page info extracted feels excessive.

That's using the "Extract and cache book information" toggle to refresh the whole documents directory (not recursively).

For comparison, using the simulator on a list of 31 books (including a few heavy hitters: 7 1200-7000 pages books), the timings are:
* 8 seconds for metadata only
* 35 seconds (x4) when rendering
* 19 seconds (x2) when using `estimatePageCount`

**Comment by benoit-pierre** [comment]
OK, so after removing all display updates during the extraction to get more accurate timings:

|                               | simulator (31 books) | kindle (25 books) |
|-------------------------------|----------------------|-------------------|
|metadata only                  |             4s       |         11s       |
|estimate (`estimatePageCount`) |            15s (x4)  |         44s (x4)  |
|estimate (`render`)            |            29s (x7)  |         82s (x7)  |

@poire-z: do you have a link to your full code for the size based estimate? I'd like to test it.

There's also something screwy going on depending on whether a book as already been opened or standby has been used on my Kindle. I've changed my settings to "start with file browser", and after starting Koreader, a round of "Extract and cache book information" directly after startup to the browser takes between ~60 and ~100 seconds, only going back to ~10 seconds after a brief stint into standby or having opened a book.

**Comment by poire-z** [comment]
Here's my 3 patches of my afternoon experiment: [rawtextsize1.zip](https://github.com/koreader/koreader/files/9833768/rawtextsize1.zip) (pick what you need, and depending on your progress and energy, compare the accuracy of the estimated page number vs. the rendered page number if opened, for your estimatePageCount method vs my raw_text_size method).

> directly after startup to the browser takes between ~60 and ~100 seconds, only going back to ~10 seconds after a brief stint into standby or having opened a book.

No idea. Could be that some data is read from SDCard to memory and reused from memory on next attemps, so reading data from SDCard is avoided.
Also, extractBookInfo may save cached DOMs into `koreader/cache/tmpcr3cache/`, but it should allow to keep only one item in there - so may be something is screwed and your not-that-excessive timings could be reusing caches found there?

**Comment by NiLuJe** [comment]
> There's also something screwy going on depending on whether a book as already been opened or standby has been used on my Kindle

Check in frameworkless mode?

(This *very* vaguely feels familiar, but I've hardly ever used KOReader on Kindles).

**Comment by NiLuJe** [comment]
> What does that do exactly? For that matter, do you know what the "ASAP" startup does exactly?

It mainly kills the java UI to free more RAM. But may also affect weird random Kindle crap, so, who knows ;o).
(The ASAP variant skip various workarounds to prevent the framework from bleeding through the UI on startup; this becomes a non-issue without said framework running ;p)

**Comment by benoit-pierre** [comment]
> > What does that do exactly? For that matter, do you know what the "ASAP" startup does exactly?
> 
> It mainly kills the java UI to free more RAM. But may also affect weird random Kindle crap, so, who knows ;o).

Ah! I guess that explains why the kindle then reboot when exiting Koreader ;)

> (The ASAP variants skip various workarounds that slow down the startup to prevent the framework from bleeding through the UI on startup; this becomes a non-issue without said framework running ;p)

I see, thanks.

**Comment by poire-z** [comment]
@benoit-pierre : I think you followed up _somewhere_ on your experiments' results - but I can't find that somewhere :/ (cross refs are a good thing :)
I had kept notes of your branches https://github.com/koreader/koreader/compare/master...benoit-pierre:koreader:bpierre and https://github.com/koreader/koreader/compare/master...benoit-pierre:koreader:bpierre with your work in progress, so you must have mentionned them _somewhere_.

Do you remember what were your results and conclusions? Have you given up on the idea, and why?, or is it still in your plans to continue with it ?
I might be willing to go with my "file size" solution, just to get something to show as pages for non-opened books. I don't know/remember if your other more complex solutions really gave a better (or less wrong) page count estimation.

**Comment by benoit-pierre** [comment]
@poire-z: here's an updated [branch](https://github.com/benoit-pierre/koreader/compare/meson+monorepo-ish...benoit-pierre:koreader:book_length).

> Do you remember what were your results and conclusions? 

OK, so I went trough a collection of 971 books, selecting 14 of them as reference. I opened all of them in the simulator do get the number of pages after rendering (with a few style tweaks to reduce differences between books). I then calculated the `bytes per synthetic page` and `chars per synthetic page` values from those 14 reference books and then the estimated number of pages for all books using both methods.

The text size method tends to overestimate the number of pages:
- min difference with the rendering pages: -33%
- max difference: ×2.5 max
- on average +2%
- underestimate by -20% or more: 19 times (~-24%)
- overestimate by 20% or more: 81 times (~+40%)

The text length method rather tends to underestimate:
- min difference: ÷2.2
- max difference: +20%
- on average -1%
- underestimate by -20% or more: 21 times (~-32%)
- overestimate by 20% or more: 1 time

But the big difference is when the text length method gives a lower or bigger estimate, it makes sense: e.g. a play or script (very sparse page layout) will get a lower estimated count.

As for performance, the text length method is obviously much slower, for the whole series of 971 books on PC:
- the text size method is 7% slower
- the text length method is ×4.2 slower (with my [optimized version](https://github.com/benoit-pierre/koreader/blob/73c59b7ccab63b9790f14f53bf7173376162e0d0/subprojects/crengine/crengine/src/epubfmt.cpp#L1118), instead of a [derived `LVXMLParserCallback` class](https://github.com/benoit-pierre/koreader/blob/73c59b7ccab63b9790f14f53bf7173376162e0d0/subprojects/crengine/crengine/src/epubfmt.cpp#L1086), which is ×11.7 slower)

> Have you given up on the idea, and why?, or is it still in your plans to continue with it ?

I use it, it works fine.

Done:
- estimating the number of pages is optional (no performance impact if not enabled)
- showing the length/progress dots is optional (so is showing the file type/size and showing the textual progress information)
- the `chars per synthetic page` value is configurable (with an option to calculate it from past opened books in the settings dialog), but not dynamically updated (I don't want the length to change as I open new books)
- for the same reason, the estimated length is always used over the SDR information (doubling the font size and the number of rendered pages should not double the estimated length of the book)
- when showing the estimated number pages, it's rounded similarly to what you had proposed

Missing:
- calculating the raw text length is only supported for EPUBs (and only UTF-8 is supported)
- there are potential issues with my changes to the way the layout is done in `plugins/coverbrowser.koplugin/listmenu.lua`

**Comment by poire-z** [comment]
In your crengine data extraction, you are still using the old (really slower) way to lookup files:
```c
        for ( size_t i=0; i<spineItemsNb; i++ ) {
            if (!spineItems[i]->is_xhtml)
                continue;
            lString32 name = LVCombinePaths(codeBase, spineItems[i]->href);
            for ( size_t j=0; j<nb_files; j++ ) {
                const LVContainerItemInfo * item = m_arc->GetObjectInfo(j);
```
Since https://github.com/koreader/crengine/pull/495 and its first commit, you don't need that 2nd loop, and I think you can use `m_arc->GetObjectInfo(name);`.
On books with many files, I think it could get you a noticable speedup.

**Comment by poire-z** [comment]
I see that in your frontend code, you just went with 16 (optimized parsing of all xhtml files) - and removed all options (which is fine by me, in the end, we should settle on a single way to do it).
What about 4 vs 8 (non-optimized parsing, full vs partial) ? Is the "partial" good enough or not? Would a faster 16+8 (partial optimized) be of interest?

> estimating the number of pages is optional (no performance impact if not enabled)

Ok with that (but if we get something fast, we may always do it - so that toggling the settings don't need to refresh the bookinfo caches).

> showing the length/progress dots is optional 

Progress dots would be better done in another PR (if you decide to make PRs :), so we can discuss all that and the layout issues independantly.
Moreover, we may (or not) decide to make bookinfo.sdb part of core (https://github.com/koreader/koreader/issues/8472#issuecomment-1554638967), so coverbrowser settings will probably go being stored in our settings.reader.lua.

> the chars per synthetic page value is configurable (with an option to calculate it from past opened books in the settings dialog), but not dynamically updated (I don't want the length to change as I open new books)

I initially (when you started that) was against the idea of having the user decide for a "chars per page" value, and I would rather want that to be automatic so he doesn't have to wonder - or us to go with an arbitrary default of 1024.
But I like the idea of this added extra button `Calculate from already analyzed and past opened books` (which could also be a menu item of its own), so I guess I may be fine with not updating the page SQL column and not having sqlite compute the estimated page (as I proposed).
I guess I'm then ok with your _I don't want the length to change as I open new books_ (although with the rounding, this may rarely be noticable?)

> for the same reason, the estimated length is always used over the SDR information (doubling the font size and the number of rendered pages should not double the estimated length of the book)

If you mean that if for one book I have opened and am reading, I change the font size to be bigger than the others and get a real page number of 200, you would still prefer to show the estimated page number of 100 ?! I'd rather see the real page number picked from the SDR.

**Comment by poire-z** [comment]
> I use it, it works fine.

Can you tell a bit about your plans, with this little "estimated page number" feature, and the bigger https://github.com/koreader/koreader-base/pull/1586 updates & build changes ?

**A** Are you just happy with having all that as a fork and building for yourself, and keeping it updated with our normal changes?
**B** Or do you plan on sometime making PRs and getting into heavy discussions with us and compromises ? :)
**C** ??

I (we?) am not forcing you into any way, it's all your choice.
I'm actually quite afraid of **B** :) and a little sad with **A** :)
I could pick some nice stuff from your **A**, but then, it will be complicated for you to do the sync each time I do that.

If you decide to go with **B**, you could followup in https://github.com/koreader/koreader-base/pull/1586 and give some estimate of an appropriate time we could all work on it (soon it is summer, I'll probably be available most of the time, but @Frenzie @NiLuJe may not be). And/or quickly discuss and decide if/how we want it.

Also, your "estimated page number" is based on your "meson" reorg, so we can't pick it as-is, and depending on the ordering of the works/PRs, you may need to rebase it on our current code. And there's this bookinfo sqlite db move that may block/be blocked by these works.

**Comment by benoit-pierre** [comment]
> In your crengine data extraction, you are still using the old (really slower) way to lookup files:
> 
> […]
> 
> Since [koreader/crengine#495](https://github.com/koreader/crengine/pull/495) and its first commit, you don't need that 2nd loop, and I think you can use `m_arc->GetObjectInfo(name);`. On books with many files, I think it could get you a noticable speedup.

Right, I amended the code (but that's not used by the "optimized raw text length" path).

Updated timings:


| Method  | minimum | maximum | average | total |
| ------- | ------- | ------- | ------- | ----- |
| minimal | 1 ms | 13 ms | 2 ms | 1.673 s |
| raw text size | 1 ms | 13 ms | 2 ms | 1.733 s (+4%) |
| raw text length (reference) | 2 ms | 284 ms | 20 ms | 19.347 s (×11.6) |
| raw text length (optimized) | 1 ms | 83 ms | 7 ms | 6.908 s (×4.1) |

And I also run the benchmark on my Kindle with a more limited collection of 64 books:

| Method  | minimum | maximum | average | total |
| ------- | ------- | ------- | ------- | ----- |
| minimal | 30 ms | 125 ms | 39 ms | 2.462 s |
| raw text size | 31 ms | 146 ms | 41 ms | 2.578 s (+5%) |
| raw text length (reference) | 37 ms | 3.032 s | 346 ms |  22.013 (×8.9) |
| raw text length (optimized) | 33 ms | 641 ms | 107 ms | 6.770 (×2.7) |

**Comment by benoit-pierre** [comment]
> > for the same reason, the estimated length is always used over the SDR information (doubling the font size and the number of rendered pages should not double the estimated length of the book)
> 
> If you mean that if for one book I have opened and am reading, I change the font size to be bigger than the others and get a real page number of 200, you would still prefer to show the estimated page number of 100 ?! I'd rather see the real page number picked from the SDR.

Just to be clear: just for calculating the number of dots.

**Comment by benoit-pierre** [comment]
> I see that in your frontend code, you just went with 16 (optimized parsing of all xhtml files) - and removed all options (which is fine by me, in the end, we should settle on a single way to do it). What about 4 vs 8 (non-optimized parsing, full vs partial) ? Is the "partial" good enough or not?

It suffers from the same issues as the "raw text size" method: overestimating the length of style heavy books, which is why I've not bothered with an optimized variant.

**Comment by Ceynou** [comment]
Would someone be kind enough to add a basic calculation option in the meantime a better solution is found?
Something like rendering every book without having to open them, doesn't matter if it takes like an hour to process them all

**Comment by Commodore64user** [comment]
> > I use it, it works fine.
> 
> Can you tell a bit about your plans, with this little "estimated page number" feature, and the bigger [koreader/koreader-base#1586](https://github.com/koreader/koreader-base/pull/1586) updates & build changes ?
> 
> **A** Are you just happy with having all that as a fork and building for yourself, and keeping it updated with our normal changes? **B** Or do you plan on sometime making PRs and getting into heavy discussions with us and compromises ? :) **C** ??
> 
> I (we?) am not forcing you into any way, it's all your choice. I'm actually quite afraid of **B** :) and a little sad with **A** :) I could pick some nice stuff from your **A**, but then, it will be complicated for you to do the sync each time I do that.
> 
> If you decide to go with **B**, you could followup in [koreader/koreader-base#1586](https://github.com/koreader/koreader-base/pull/1586) and give some estimate of an appropriate time we could all work on it (soon it is summer, I'll probably be available most of the time, but @Frenzie @NiLuJe may not be). And/or quickly discuss and decide if/how we want it.
> 
> Also, your "estimated page number" is based on your "meson" reorg, so we can't pick it as-is, and depending on the ordering of the works/PRs, you may need to rebase it on our current code. And there's this bookinfo sqlite db move that may block/be blocked by these works.

We are all here on tenterhooks as @benoit-pierre is yet to answer this question.

**Comment by benoit-pierre** [comment]
I have no plan to PR that "estimated page number" feature, if that's what you're asking.

**Comment by Philantrop** [comment]
@benoit-pierre, since the "estimated page number" feature is important to me as a reader, I'd like to try and cherry-pick the relevant commits from the "bpierre" branch of your repository. Could you possibly kindly direct me to the relevant commits (e. g. as in https://github.com/koreader/koreader/issues/8390#issuecomment-1285755184)? 

And thank you for working on this.

**Comment by benoit-pierre** [comment]
@Philantrop: you'll need:
- [document: don't strip extra properties](https://github.com/benoit-pierre/koreader/commit/2e0f57d5dd294ce2ac848358d2cf62e1972552f9)
- [crengine: support calculating the raw text size / length](https://github.com/benoit-pierre/koreader/commit/dfc969caf03464fb38b528669f4eedddb4205fba)
- [coverbrowser: simplify detailed mode layout](https://github.com/benoit-pierre/koreader/commit/cf8b505a5d84056d29212bc05ceed09c8ccc327e)
- [coverbrowser: add progress / length dots + page count estimate](https://github.com/benoit-pierre/koreader/commit/fb3a04175fa4c7223ea1c71598e2b6253b977454)

**Comment by joshuacant** [comment]
I'm not sure if this is the right place or that my comment will be welcome here, but I have no plans to submit a PR or fork and release my own version of the coverbrowser plugin, so this might be the best (or only) place to document how I solved this for myself. It may be of use to someone, or this comment may be a bother, in which case I apologize.

I use the "Count Pages" plugin for Calibre to do the page count calculation. I do this before sending the files to my device with koreader installed, and I have Calibre configured to add the page count number to the filename. It does so by automatically adding 'P(999)' when sending the file to my reader, where 999 is whatever the actual calculated page count is.

I modified 'listmenu.lua' inside the 'coverbrowser.koplugin' folder, adding the following code above the line `-- Build the right widget` which is at line 417 in the current version of the file (as of this post).

```
            local fn_page_count = string.match(filename_without_suffix, "P%((%d+)%)")
            if fn_page_count then
                local unread_chr = "◽"
                local read_chr = "◾"
                local fn_pages = tonumber(fn_page_count)

                -- 125 pages per 1cm in avg hardcover
                -- 5 ◽ per cm on 16x12cm eINK display
                -- thus 25 pages per ◽
                -- maximum of 30 blocks (750 pages) to leave room for title
                -- minimum of 2 blocks otherwise book goes from ◽ to ◾ immediately
                local total_blocks = math.max((math.min(math.floor((fn_pages/25)+0.5), 30)), 2)

                if status == "complete" or status == "abandoned" then
                    pages_str = string.rep(read_chr, total_blocks)
                elseif percent_finished then
                    local finished_blocks = math.max(math.floor(total_blocks * percent_finished), 1)
                    local unfinished_blocks = total_blocks - finished_blocks
                    pages_str = string.rep(read_chr, finished_blocks) .. string.rep(unread_chr, unfinished_blocks)
                else
                    pages_str = string.rep(unread_chr, total_blocks)
                end
            end
```

This works, for me, on my device (Kobo Sage), with my books. It also (because I'm a crazy person) generates indicators that approximate the actual real-life thickness of hardcover books (up to a limit of 750 pages). I feel certain there are a whole bunch of edge cases that I have not accounted for, but I made this for an audience of one. I prefer this over on-device page calculations (adjustable "accuracy", speed) but I do think it would be nice if koreader had something similar built-in. Being able to quickly size up a selection of potential reads and matching book size to appetite is an essential part of choosing my next book. It must be for others as well.

![FileManager_2024-07-15_173709](https://github.com/user-attachments/assets/54148a21-81e6-4ae7-953b-67bd4649e9f0)

**Comment by fer1803toledo** [comment]
Excuse me, but I'm very clumsy. I am very interested in this function. It is now possible to mark the length of the book without opening the file. If so...would someone be so kind as to indicate how it can be installed? I tried to copy the files above and the koreader installation was loaded...thanks in advance.
